### PR TITLE
Make encrypted volatile volume optional

### PR DESCRIPTION
**Pull request by marmarek**
Make 'ephemeral' volume property accessible read-write via Admin API, and set
the default value to False. This allows the feature to be enabled for specific
VMs where it is needed. This can be changed via `qvm-volume` tool.
Furthermore, volumes of a new DispVM gets the value from its template's
matching volumes. Do the same for 'rw' property.

QubesOS/qubes-issues#904

### CODE REVIEW

File: qubes/storage/__init__.py [hunk 1]
```diff
@@ -162,10 +166,25 @@ def __xml__(self):
...
+
+    @ephemeral.setter
+    def ephemeral(self, value):
+        if not value:
+            self._ephemeral = False
+            return
+        if self.snap_on_start or self.save_on_stop or \
+                self.domain is not None or not self.rw:
+            raise ValueError(
+                'Cannot enable ephemeral on snap_on_star or save_on_stop or '
```
Comment (DemiMarie) [inline qubes/storage/__init__.py:185]:
```suggestion
                'Cannot enable ephemeral on snap_on_start or save_on_stop or '
```

File: qubes/storage/__init__.py [hunk 2]
```diff
@@ -162,10 +166,25 @@ def __xml__(self):
...
+        # default value
+        return False
+
+    @ephemeral.setter
+    def ephemeral(self, value):
+        if not value:
+            self._ephemeral = False
+            return
+        if self.snap_on_start or self.save_on_stop or \
+                self.domain is not None or not self.rw:
```
Comment (DemiMarie) [inline qubes/storage/__init__.py:183]: Will that prevent DispVM templates from starting?

File: qubes/tests/vm/dispvm.py [hunk 3]
```diff
@@ -320,3 +321,24 @@ def test_022_storage_app_change(self):
...
+                as mock_domains:
+            mock_domains.configure_mock(**{
+                'get_new_unused_dispid': mock.Mock(return_value=42),
+                '__getitem__.side_effect': orig_getitem
+            })
+            dispvm = self.app.add_new_vm(qubes.vm.dispvm.DispVM,
+                name='test-dispvm', template=self.appvm)
+            self.loop.run_until_complete(dispvm.create_on_disk())
+        self.assertIs(dispvm.template, self.appvm)
+        self.assertTrue(dispvm.volumes['volatile'].ephemeral)
```
Comment (DemiMarie) [inline qubes/tests/vm/dispvm.py:344]:
```suggestion
        self.assertTrue(dispvm.volumes['volatile'].ephemeral)
```

File: qubes/storage/__init__.py (hunk 2 above)
Comment (marmarek) [reply qubes/storage/__init__.py:183]: It should not. It's about `volatile` volume only, which (currently) always have the same properties regardless of qube type. Furthermore, this is done on setting the property, not starting the qube.

Comment (DemiMarie) [reply qubes/storage/__init__.py:183]: Ah, good catch :woman_facepalming:

File: qubes/tests/vm/dispvm.py (hunk 3 above)
Comment (marmarek) [reply qubes/tests/vm/dispvm.py:344]: Where is the difference?

Comment (DemiMarie) [reply qubes/tests/vm/dispvm.py:344]: Missing newline at EOF

**Comment by marmarek** [comment]
> Instead of the `ephemeral` property for volatile volumes defaulting to `False` unconditionally, I strongly believe this should be configurable system-wide via `qubes-prefs` or the Qubes Global Settings GUI. 

I don't exclude adding such option later. But as discussed before, the use case (besides DispVM) is rather niche, so I'm not going to spend too much time on this now (delaying R4.1 even more).

> It should also be configurable on TemplateVMs.

For DispVMs, I specifically added inheriting the setting from their templates. For others: see above.

**Pull request merged**